File: ObjcExample/ObjcExample/Demo/PinLocation/GKPinLocationViewController.m
//
//  GKPinLocationViewController.m
//  ObjcExample
//
//  Created by gaokun on 2021/2/5.
//

#import "GKPinLocationViewController.h"
#import "JXCategoryPinTitleView.h"
#import <GKPageSmoothView/GKPageSmoothView.h>
#import "GKPinLocationView.h"

@interface GKPinLocationViewController ()<GKPageSmoothViewDataSource, GKPageSmoothViewDelegate, JXCategoryViewDelegate>

@property (nonatomic, strong) GKPageSmoothView *smoothView;

@property (nonatomic, strong) UIImageView *headerView;

@property (nonatomic, strong) JXCategoryPinTitleView *titleView;

@end

@implementation GKPinLocationViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.gk_navTitle = @"上下滑动切换tab";
    self.gk_navLineHidden = YES;
    self.gk_navTitleColor = [UIColor whiteColor];
    self.gk_navTitleFont = [UIFont boldSystemFontOfSize:18.0f];
    self.gk_navBackgroundColor = [UIColor clearColor];
    self.gk_statusBarStyle = UIStatusBarStyleLightContent;
    UIBarButtonItem *oriItem = [UIBarButtonItem gk_itemWithTitle:@"原点" target:self action:@selector(oriAction)];
    UIBarButtonItem *criItem = [UIBarButtonItem gk_itemWithTitle:@"临界点" target:self action:@selector(criAction)];
    self.gk_navRightBarButtonItems = @[oriItem, criItem];
    
    [self.view addSubview:self.smoothView];
    [self.smoothView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.edges.equalTo(self.view);
    }];
    
    [self.smoothView reloadData];
}

- (void)oriAction {
    [self.smoothView scrollToOriginalPoint];
}

- (void)criAction {
    [self.smoothView scrollToCriticalPoint];
}

#pragma mark - GKPageSmoothViewDataSource
- (UIView *)headerViewInSmoothView:(GKPageSmoothView *)smoothView {
    return self.headerView;
}

- (UIView *)segmentedViewInSmoothView:(GKPageSmoothView *)smoothView {
    return self.titleView;
}

- (NSInteger)numberOfListsInSmoothView:(GKPageSmoothView *)smoothView {
    return 1;
}

- (id<GKPageSmoothListViewDelegate>)smoothView:(GKPageSmoothView *)smoothView initListAtIndex:(NSInteger)index {
    GKPinLocationView *listView = [[GKPinLocationView alloc] init];
    
    NSMutableArray *data = [NSMutableArray new];
    NSArray *counts = @[@(6), @(8), @(9), @(5), @(7), @(10), @(13), @(6), @(8)];
    for (int i = 0; i < self.titleView.titles.count; i++) {
        NSMutableDictionary *dic = [NSMutableDictionary new];
        dic[@"title"] = self.titleView.titles[i];
        dic[@"count"] = counts[i];
        [data addObject:dic];
    }
    listView.data = data;
    return listView;
}

#pragma mark - GKPageSmoothViewDelegate
- (void)smoothView:(GKPageSmoothView *)smoothView listScrollViewDidScroll:(UIScrollView *)scrollView contentOffset:(CGPoint)contentOffset {
    if (!(scrollView.isTracking || scrollView.isDecelerating)) return;
    
    //用户滚动的才处理
    //获取categoryView下面一点的所有布局信息，用于知道，当前最上方是显示的哪个section
    CGFloat categoryH = self.titleView.frame.size.height;
    
    UITableView *tableView = (UITableView *)scrollView;
    NSArray <NSIndexPath *> *topIndexPaths = [tableView indexPathsForRowsInRect:CGRectMake(0, contentOffset.y + categoryH - self.headerView.frame.size.height + 40 + 10, tableView.frame.size.width, 200)];
    
    NSIndexPath *topIndexPath = topIndexPaths.firstObject;
    NSUInteger topSection = topIndexPath.section;
    if (topIndexPath != nil) {
        if (self.titleView.selectedIndex != topSection) {
            [self.titleView selectItemAtIndex:topSection];
        }
    }
}

#pragma mark - JXCategoryViewDelegate
- (void)categoryView:(JXCategoryBaseView *)categoryView didClickSelectedItemAtIndex:(NSInteger)index {
    UITableView *tableView = (UITableView *)self.smoothView.currentListScrollView;
    CGRect frame = [tableView rectForHeaderInSection:index];
    [tableView setContentOffset:CGPointMake(0, frame.origin.y - kBaseHeaderHeight + kBaseSegmentHeight + 40) animated:YES];
}

#pragma mark - 懒加载
- (GKPageSmoothView *)smoothView {
    if (!_smoothView) {
        _smoothView = [[GKPageSmoothView alloc] initWithDataSource:self];
        _smoothView.ceilPointHeight = GK_STATUSBAR_NAVBAR_HEIGHT;
        _smoothView.delegate = self;
    }
    return _smoothView;
}

- (UIImageView *)headerView {
    if (!_headerView) {
        _headerView = [[UIImageView alloc] initWithFrame:CGRectMake(0, 0, kScreenW, kBaseHeaderHeight)];
        _headerView.contentMode = UIViewContentModeScaleAspectFill;
        _headerView.clipsToBounds = YES;
        _headerView.image = [UIImage imageNamed:@"test"];
    }
    return _headerView;
}

- (JXCategoryPinTitleView *)titleView {
    if (!_titleView) {
        _titleView = [JXCategoryPinTitleView new];
        _titleView.backgroundColor = UIColor.whiteColor;
        _titleView.frame = CGRectMake(0, 0, kScreenW, kBaseSegmentHeight);
        _titleView.titles = @[@"年货市集", @"新年换新", @"安心过年", @"爆品专区", @"尝鲜专区", @"贺岁大餐", @"超值外卖", @"玩乐特惠", @"商超年货"];
        _titleView.pinImage = [UIImage gk_changeImage:[UIImage imageNamed:@"location"] color:UIColor.redColor];
        _titleView.titleColor = UIColor.grayColor;
        _titleView.titleSelectedColor = UIColor.redColor;
        _titleView.titleFont = [UIFont systemFontOfSize:15];
        _titleView.delegate = self;
    }
    return _titleView;
}

@end

File: SwiftExample/SwiftExample/Demo/PinLocation/GKPinLocationViewController.swift
//
//  GKPinLocationViewController.swift
//  SwiftExample
//
//  Created by gaokun on 2021/3/18.
//

import UIKit
import GKNavigationBarSwift
import JXSegmentedView
import GKPageSmoothView

class GKPinLocationViewController: GKDemoBaseViewController {
    lazy var smoothView: GKPageSmoothView = {
        let smoothView = GKPageSmoothView(dataSource: self)
        smoothView.ceilPointHeight = GK_STATUSBAR_NAVBAR_HEIGHT
        smoothView.delegate = self
        return smoothView
    }()
    
    lazy var headerView: UIImageView = {
        let headerView = UIImageView(frame: CGRect(x: 0, y: 0, width: kScreenW, height: kBaseHeaderHeight))
        headerView.contentMode = .scaleAspectFill;
        headerView.clipsToBounds = true;
        headerView.image = UIImage(named: "test")
        return headerView
    }()
    
    let pinDataSource = JXSegmentedPinTitleDataSource()
    lazy var titleView: JXSegmentedView = {
        pinDataSource.titles = ["年货市集", "新年换新", "安心过年", "爆品专区", "尝鲜专区", "贺岁大餐", "超值外卖", "玩乐特惠", "商超年货"];
        pinDataSource.pinImage = self.changeImage(image: UIImage(named: "location")!, color: UIColor.red)
        pinDataSource.titleNormalColor = UIColor.gray
        pinDataSource.titleSelectedColor = UIColor.red
        pinDataSource.titleNormalFont = UIFont.systemFont(ofSize: 15)
    
        let titleView = JXSegmentedView()
        titleView.backgroundColor = UIColor.white
        titleView.frame = CGRect(x: 0, y: 0, width: kScreenW, height: kBaseSegmentHeight)
        titleView.delegate = self
        titleView.dataSource = pinDataSource
        return titleView
    }()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.gk_navTitle = "上下滑动切换tab"
        self.gk_navLineHidden = true
        self.gk_navTitleColor = .white
        self.gk_navTitleFont = UIFont.boldSystemFont(ofSize: 18)
        self.gk_navBackgroundColor = .clear
        self.gk_statusBarStyle = .lightContent
        
        view.addSubview(smoothView)
        smoothView.snp.makeConstraints { (make) in
            make.edges.equalTo(self.view)
        }
        smoothView.reloadData()
    }
    
    func changeImage(image: UIImage, color: UIColor) -> UIImage {
        UIGraphicsBeginImageContextWithOptions(image.size, false, image.scale)
        let context = UIGraphicsGetCurrentContext()
        context?.translateBy(x: 0, y: image.size.height)
        context?.scaleBy(x: 1.0, y: -1.0)//kCGBlendModeNormal
        context?.setBlendMode(.normal)
        let rect = CGRect(x: 0, y: 0, width: image.size.width, height: image.size.height)
        context?.clip(to: rect, mask: image.cgImage!);
        color.setFill()
        context?.fill(rect)
        let newImage = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()
        return newImage!
    }
}

extension GKPinLocationViewController: GKPageSmoothViewDataSource, GKPageSmoothViewDelegate {
    func headerView(in smoothView: GKPageSmoothView) -> UIView {
        return self.headerView
    }
    
    func segmentedView(in smoothView: GKPageSmoothView) -> UIView {
        return self.titleView
    }
    
    func numberOfLists(in smoothView: GKPageSmoothView) -> Int {
        return 1
    }
    
    func smoothView(_ smoothView: GKPageSmoothView, initListAtIndex index: Int) -> GKPageSmoothListViewDelegate {
        let listView = GKPinLocationView()
        var data = [[String: String]]()
        let counts = [6, 8 ,9, 5, 7, 10, 13, 6, 8]
        for i in 0..<counts.count {
            var dic = [String: String]()
            dic["title"] = self.pinDataSource.titles[i]
            dic["count"] = String(counts[i])
            data.append(dic)
        }
        listView.datas = data;
        return listView;
    }
    
    func smoothViewListScrollViewDidScroll(_ smoothView: GKPageSmoothView, scrollView: UIScrollView, contentOffset: CGPoint) {
        if !(scrollView.isTracking || scrollView.isDecelerating) { return }
        // 用户滚动的才处理
        // 获取categoryView下面一点的所有部件信息，用于指定，当前最上方显示的那个section
        let categoryH = self.titleView.frame.size.height
        let tableView = scrollView as! UITableView
        let topIndexPaths = tableView.indexPathsForRows(in: CGRect(x: 0, y: contentOffset.y + categoryH - self.headerView.frame.size.height + 40 + 10, width: tableView.frame.size.width, height: 200))
        let topIndexPath = topIndexPaths?.first
        if let indexPath = topIndexPath {
            let topSection = indexPath.section
            if self.titleView.selectedIndex != topSection {
                self.titleView.selectItemAt(index: topSection)
            }
        }
    }
}

extension GKPinLocationViewController: JXSegmentedViewDelegate {
    func segmentedView(_ segmentedView: JXSegmentedView, didClickSelectedItemAt index: Int) {
        let tableView = self.smoothView.currentListScrollView as! UITableView
        let frame = tableView.rectForHeader(inSection: index)
        tableView.setContentOffset(CGPoint(x: 0, y: frame.origin.y - kBaseHeaderHeight + kBaseSegmentHeight + 40), animated: true)
    }
}
